From: Michael Ploujnikov <michael.ploujnikov@oracle.com>
Date: Mon, 4 Nov 2013 11:37:12 +0100
Subject: qla2xxx: Module parameter 'ql2xasynclogin'
References: bnc#825896
Patch-Mainline: Not yet

Add module parameter 'ql2xasynclogin' to control use of
sync or sync port login.


Signed-off-by: Michael Ploujnikov <michael.ploujnikov@oracle.com>
Signed-off-by: Hannes Reinecke <hare@suse.de>

diff --git a/drivers/scsi/qla2xxx/qla_gbl.h b/drivers/scsi/qla2xxx/qla_gbl.h
index 36f3c9f..de4604b 100644
--- a/drivers/scsi/qla2xxx/qla_gbl.h
+++ b/drivers/scsi/qla2xxx/qla_gbl.h
@@ -111,6 +111,7 @@ extern int ql2xdontresethba;
 extern unsigned int ql2xmaxlun;
 extern int ql2xmdcapmask;
 extern int ql2xmdenable;
+extern int ql2xasynclogin;
 
 extern int qla2x00_loop_reset(scsi_qla_host_t *);
 extern void qla2x00_abort_all_cmds(scsi_qla_host_t *, int);
diff --git a/drivers/scsi/qla2xxx/qla_init.c b/drivers/scsi/qla2xxx/qla_init.c
index 3e29799..16043ac 100644
--- a/drivers/scsi/qla2xxx/qla_init.c
+++ b/drivers/scsi/qla2xxx/qla_init.c
@@ -3669,7 +3669,7 @@ qla2x00_fabric_dev_login(scsi_qla_host_t *vha, fc_port_t *fcport,
 	rval = QLA_SUCCESS;
 	retry = 0;
 
-	if (IS_ALOGIO_CAPABLE(ha)) {
+	if (ql2xasynclogin && IS_ALOGIO_CAPABLE(ha)) {
 		if (fcport->flags & FCF_ASYNC_SENT)
 			return rval;
 		fcport->flags |= FCF_ASYNC_SENT;
diff --git a/drivers/scsi/qla2xxx/qla_os.c b/drivers/scsi/qla2xxx/qla_os.c
index 3ff8bbc..dc7f3b3 100644
--- a/drivers/scsi/qla2xxx/qla_os.c
+++ b/drivers/scsi/qla2xxx/qla_os.c
@@ -207,6 +207,13 @@ MODULE_PARM_DESC(ql2xmdenable,
 		"0 (Default) - MiniDump disabled. "
 		"1 - MiniDump enabled.");
 
+int ql2xasynclogin = 1;
+module_param(ql2xasynclogin, int, S_IRUGO);
+MODULE_PARM_DESC(ql2xasynclogin,
+		"Enables asynchronous login to remote ports "
+		"Default is 1 - Use asynchronous login if possible.");
+
+
 /*
  * SCSI host template entry points
  */
@@ -3527,7 +3534,7 @@ void qla2x00_relogin(struct scsi_qla_host *vha)
 					}
 				}
 
-				if (IS_ALOGIO_CAPABLE(ha)) {
+				if (ql2xasynclogin && IS_ALOGIO_CAPABLE(ha)) {
 					fcport->flags |= FCF_ASYNC_SENT;
 					data[0] = 0;
 					data[1] = QLA_LOGIO_LOGIN_RETRIED;
